File: stock-analyzer-java/src/main/java/com/stockanalyzer/scrapers/TickertapeScraper.java
package com.stockanalyzer.scrapers;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.stockanalyzer.models.FinancialData;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Scraper for Tickertape.in financial data.
 */
public class TickertapeScraper extends BaseScraper {

    @Override
    public FinancialData scrape(String url) {
        Document doc = fetchPage(url);
        if (doc == null) {
            logger.warn("Failed to fetch Tickertape.in page: {}", url);
            return data;
        }

        extractCompanyName(doc);
        extractOverview(doc);
        extractJsonData(doc);
        extractValuation(doc);
        extractGrowth(doc);
        extractCashFlowData(doc, url);

        return data;
    }

    private void extractCompanyName(Document doc) {
        Element nameElem = doc.selectFirst("h1[class*=security-name], h1[class*=stock-name], h1");
        if (nameElem != null && data.getCompanyName() == null) {
            data.setCompanyName(nameElem.text().trim());
        }
    }

    private void extractOverview(Document doc) {
        // Market price
        Element priceElem = doc.selectFirst("span[class*=current-price], span[class*=price]");
        if (priceElem != null) {
            Double price = cleanNumber(priceElem.text());
            if (price != null) {
                data.addRatio("Current Price", price);
            }
        }

        // Extract metrics from cards/sections
        Elements metricCards = doc.select("div[class*=metric], div[class*=stat], span[class*=info-item]");
        for (Element card : metricCards) {
            Element labelElem = card.selectFirst("span[class*=label], div[class*=title], span[class*=name]");
            Element valueElem = card.selectFirst("span[class*=value], div[class*=number]");
            if (labelElem != null && valueElem != null) {
                String label = labelElem.text().trim();
                Double value = cleanNumber(valueElem.text());
                if (value != null) {
                    data.addRatio(label, value);
                }
            }
        }

        // Extract from definition lists
        Elements dlElements = doc.select("dl");
        for (Element dl : dlElements) {
            Elements dts = dl.select("dt");
            Elements dds = dl.select("dd");
            for (int i = 0; i < Math.min(dts.size(), dds.size()); i++) {
                String label = dts.get(i).text().trim();
                Double value = cleanNumber(dds.get(i).text());
                if (value != null) {
                    data.addRatio(label, value);
                }
            }
        }
    }

    private void extractJsonData(Document doc) {
        // Look for JSON data in script tags
        Elements scripts = doc.select("script[type=application/json], script");
        for (Element script : scripts) {
            String content = script.data();
            if (content == null || content.isEmpty()) continue;

            // Try to find __INITIAL_STATE__ or similar
            Pattern pattern = Pattern.compile("window\\.__INITIAL_STATE__\\s*=\\s*(\\{.*?\\});", Pattern.DOTALL);
            Matcher matcher = pattern.matcher(content);
            if (matcher.find()) {
                try {
                    parseJsonFinancialData(JsonParser.parseString(matcher.group(1)).getAsJsonObject());
                } catch (Exception e) {
                    logger.debug("Failed to parse JSON state: {}", e.getMessage());
                }
            }

            // Also try parsing direct JSON
            if (content.trim().startsWith("{")) {
                try {
                    parseJsonFinancialData(JsonParser.parseString(content).getAsJsonObject());
                } catch (Exception e) {
                    // Not valid JSON, skip
                }
            }
        }
    }

    private void parseJsonFinancialData(JsonObject jsonData) {
        if (jsonData == null) return;

        // Look for cash flow data
        if (jsonData.has("cashFlow")) {
            extractFcfFromJson(jsonData.getAsJsonObject("cashFlow"));
        }
        if (jsonData.has("cash_flow")) {
            extractFcfFromJson(jsonData.getAsJsonObject("cash_flow"));
        }

        // Look for ratios
        if (jsonData.has("ratios")) {
            JsonObject ratios = jsonData.getAsJsonObject("ratios");
            for (String key : ratios.keySet()) {
                JsonElement elem = ratios.get(key);
                if (elem.isJsonPrimitive() && elem.getAsJsonPrimitive().isNumber()) {
                    data.addRatio(key, elem.getAsDouble());
                }
            }
        }

        // Recursively search nested objects
        for (String key : jsonData.keySet()) {
            JsonElement elem = jsonData.get(key);
            if (elem.isJsonObject()) {
                parseJsonFinancialData(elem.getAsJsonObject());
            } else if (elem.isJsonArray()) {
                JsonArray arr = elem.getAsJsonArray();
                for (JsonElement item : arr) {
                    if (item.isJsonObject()) {
                        parseJsonFinancialData(item.getAsJsonObject());
                    }
                }
            }
        }
    }

    private void extractFcfFromJson(JsonObject cfData) {
        if (cfData == null) return;

        Map<String, Double> fcfValues = new LinkedHashMap<>();

        // Look for yearly FCF data
        if (cfData.has("yearly")) {
            JsonArray yearly = cfData.getAsJsonArray("yearly");
            for (JsonElement elem : yearly) {
                if (elem.isJsonObject()) {
                    JsonObject yearData = elem.getAsJsonObject();
                    String year = yearData.has("year") ? yearData.get("year").getAsString() : null;
                    Double fcf = null;
                    if (yearData.has("fcf")) {
                        fcf = yearData.get("fcf").getAsDouble();
                    } else if (yearData.has("freeCashFlow")) {
                        fcf = yearData.get("freeCashFlow").getAsDouble();
                    }
                    if (year != null && fcf != null) {
                        fcfValues.put(year, fcf);
                    }
                }
            }
        }

        if (!fcfValues.isEmpty()) {
            data.addCashFlowMetric("Free Cash Flow", fcfValues);
        }
    }

    private void extractValuation(Document doc) {
        Map<String, Pattern> patterns = Map.of(
                "PE Ratio", Pattern.compile("P/E|PE\\s*Ratio", Pattern.CASE_INSENSITIVE),
                "PB Ratio", Pattern.compile("P/B|PB\\s*Ratio", Pattern.CASE_INSENSITIVE),
                "PEG Ratio", Pattern.compile("PEG", Pattern.CASE_INSENSITIVE),
                "EV/EBITDA", Pattern.compile("EV/EBITDA", Pattern.CASE_INSENSITIVE),
                "Dividend Yield", Pattern.compile("Dividend\\s*Yield", Pattern.CASE_INSENSITIVE)
        );

        for (Map.Entry<String, Pattern> entry : patterns.entrySet()) {
            Elements elements = doc.getElementsMatchingOwnText(entry.getValue());
            for (Element elem : elements) {
                Element parent = elem.parent();
                if (parent != null) {
                    Element valueElem = parent.selectFirst("span[class*=value], div[class*=number]");
                    if (valueElem != null) {
                        Double value = cleanNumber(valueElem.text());
                        if (value != null) {
                            data.addRatio(entry.getKey(), value);
                            break;
                        }
                    }
                }
            }
        }

        // Also check tables
        Elements tables = doc.select("table");
        for (Element table : tables) {
            for (Element row : table.select("tr")) {
                Elements cells = row.select("td, th");
                if (cells.size() >= 2) {
                    String label = cells.get(0).text().trim();
                    for (Map.Entry<String, Pattern> entry : patterns.entrySet()) {
                        if (entry.getValue().matcher(label).find()) {
                            Double value = cleanNumber(cells.get(1).text());
                            if (value != null) {
                                data.addRatio(entry.getKey(), value);
                            }
                        }
                    }
                }
            }
        }
    }

    private void extractGrowth(Document doc) {
        Map<String, Pattern> patterns = Map.of(
                "Revenue Growth", Pattern.compile("Revenue\\s*Growth|Sales\\s*Growth", Pattern.CASE_INSENSITIVE),
                "Profit Growth", Pattern.compile("Profit\\s*Growth|PAT\\s*Growth", Pattern.CASE_INSENSITIVE),
                "EPS Growth", Pattern.compile("EPS\\s*Growth", Pattern.CASE_INSENSITIVE),
                "3Y Revenue CAGR", Pattern.compile("3.*Revenue.*CAGR|Revenue.*3.*CAGR", Pattern.CASE_INSENSITIVE),
                "5Y Revenue CAGR", Pattern.compile("5.*Revenue.*CAGR|Revenue.*5.*CAGR", Pattern.CASE_INSENSITIVE)
        );

        for (Map.Entry<String, Pattern> entry : patterns.entrySet()) {
            Elements elements = doc.getElementsMatchingOwnText(entry.getValue());
            for (Element elem : elements) {
                Element parent = elem.parent();
                if (parent != null) {
                    Element valueElem = parent.selectFirst("span[class*=value], div[class*=number]");
                    if (valueElem != null) {
                        Double value = cleanNumber(valueElem.text());
                        if (value != null) {
                            data.addRatio(entry.getKey(), value);
                            break;
                        }
                    }
                }
            }
        }
    }

    private void extractCashFlowData(Document doc, String baseUrl) {
        // Try to find cash flow section on main page
        Element cfSection = doc.selectFirst("section[id*=cash-flow], div[id*=cash-flow]");
        if (cfSection != null) {
            parseCashFlowTable(cfSection);
            return;
        }

        // Try to fetch cash flow sub-page
        if (baseUrl.contains("/stocks/")) {
            String cfUrl = baseUrl.replaceAll("/$", "") + "/financials?checklist=basic&statement=cashflow";
            Document cfDoc = fetchPage(cfUrl);
            if (cfDoc != null) {
                parseCashFlowPage(cfDoc);
            }
        }
    }

    private void parseCashFlowTable(Element section) {
        Element table = section.selectFirst("table");
        if (table == null) return;

        for (Element row : table.select("tr")) {
            Elements cells = row.select("td, th");
            if (cells.isEmpty()) continue;

            String metric = cells.first().text().toLowerCase();
            if (metric.contains("free cash flow") || metric.equals("fcf")) {
                Map<String, Double> fcfValues = new LinkedHashMap<>();
                for (int i = 1; i < cells.size(); i++) {
                    Double value = cleanNumber(cells.get(i).text());
                    if (value != null) {
                        fcfValues.put("Year" + i, value);
                    }
                }
                if (!fcfValues.isEmpty()) {
                    data.addCashFlowMetric("Free Cash Flow", fcfValues);
                }
            }
        }
    }

    private void parseCashFlowPage(Document doc) {
        Elements tables = doc.select("table");
        for (Element table : tables) {
            for (Element row : table.select("tr")) {
                Elements cells = row.select("td, th");
                if (cells.size() > 1) {
                    String metric = cells.first().text().toLowerCase();
                    if (metric.contains("free cash flow") || metric.equals("fcf")) {
                        Map<String, Double> fcfValues = new LinkedHashMap<>();
                        for (int i = 1; i < cells.size(); i++) {
                            Double value = cleanNumber(cells.get(i).text());
                            if (value != null) {
                                fcfValues.put("Year" + i, value);
                            }
                        }
                        if (!fcfValues.isEmpty()) {
                            data.addCashFlowMetric("Free Cash Flow", fcfValues);
                        }
                    }
                }
            }
        }
    }
}

File: stock-analyzer/scrapers/tickertape_scraper.py
"""Scraper for Tickertape.in financial data."""

from typing import Dict, Any, List, Optional
from .base_scraper import BaseScraper
import re
import json


class TickertapeScraper(BaseScraper):
    """Scraper for extracting financial data from Tickertape.in."""

    def __init__(self):
        super().__init__()
        self.company_name = ""
        self.data = {
            'overview': {},
            'financials': {},
            'cash_flow': {},
            'free_cash_flow': [],
            'valuation': {},
            'growth': {},
            'price_history': {}
        }

    def scrape(self, url: str) -> Dict[str, Any]:
        """Scrape all financial data from Tickertape.in."""
        soup = self.fetch_page(url)
        if not soup:
            return self.data

        # Extract company name
        self._extract_company_name(soup)

        # Extract overview metrics
        self._extract_overview(soup)

        # Extract financial data from embedded JSON
        self._extract_json_data(soup)

        # Extract valuation metrics
        self._extract_valuation(soup)

        # Extract growth metrics
        self._extract_growth(soup)

        # Try to get cash flow data from financials tab
        self._extract_cash_flow_data(soup, url)

        return self.data

    def _extract_company_name(self, soup) -> None:
        """Extract company name from page."""
        # Try various selectors for company name
        name_elem = soup.find('h1', class_=re.compile(r'security-name|stock-name'))
        if not name_elem:
            name_elem = soup.find('h1')
        if name_elem:
            self.company_name = name_elem.get_text(strip=True)

    def _extract_overview(self, soup) -> None:
        """Extract overview/summary metrics."""
        overview = {}

        # Market price
        price_elem = soup.find('span', class_=re.compile(r'current-price|price'))
        if price_elem:
            overview['Current Price'] = self.clean_number(price_elem.get_text(strip=True))

        # Try to find key metrics sections
        metric_cards = soup.find_all(['div', 'span'], class_=re.compile(r'metric|stat|info-item'))
        for card in metric_cards:
            label_elem = card.find(['span', 'div'], class_=re.compile(r'label|title|name'))
            value_elem = card.find(['span', 'div'], class_=re.compile(r'value|number'))
            if label_elem and value_elem:
                label = label_elem.get_text(strip=True)
                value = value_elem.get_text(strip=True)
                overview[label] = self.clean_number(value) if self._is_numeric(value) else value

        # Extract from definition lists
        dl_elements = soup.find_all('dl')
        for dl in dl_elements:
            dts = dl.find_all('dt')
            dds = dl.find_all('dd')
            for dt, dd in zip(dts, dds):
                label = dt.get_text(strip=True)
                value = dd.get_text(strip=True)
                overview[label] = self.clean_number(value) if self._is_numeric(value) else value

        self.data['overview'] = overview

    def _extract_json_data(self, soup) -> None:
        """Extract financial data from embedded JSON scripts."""
        # Look for script tags containing JSON data
        scripts = soup.find_all('script', type='application/json')
        for script in scripts:
            try:
                data = json.loads(script.string)
                self._parse_json_financial_data(data)
            except (json.JSONDecodeError, TypeError):
                continue

        # Also try to find inline JavaScript data
        scripts = soup.find_all('script')
        for script in scripts:
            if script.string:
                # Look for JSON-like structures
                json_matches = re.findall(r'window\.__INITIAL_STATE__\s*=\s*({.*?});', script.string, re.DOTALL)
                for match in json_matches:
                    try:
                        data = json.loads(match)
                        self._parse_json_financial_data(data)
                    except json.JSONDecodeError:
                        continue

    def _parse_json_financial_data(self, data: Dict) -> None:
        """Parse financial data from JSON structure."""
        if not isinstance(data, dict):
            return

        # Look for cash flow data
        if 'cashFlow' in data or 'cash_flow' in data:
            cf_data = data.get('cashFlow') or data.get('cash_flow', {})
            self._extract_fcf_from_json(cf_data)

        # Look for financial ratios
        if 'ratios' in data:
            self.data['financials'].update(data['ratios'])

        # Look for stock info
        if 'stock' in data:
            stock_data = data['stock']
            if isinstance(stock_data, dict):
                self.data['overview'].update(stock_data)

        # Recursively search for relevant data
        for key, value in data.items():
            if isinstance(value, dict):
                self._parse_json_financial_data(value)
            elif isinstance(value, list):
                for item in value:
                    if isinstance(item, dict):
                        self._parse_json_financial_data(item)

    def _extract_fcf_from_json(self, cf_data: Dict) -> None:
        """Extract Free Cash Flow data from JSON."""
        if not isinstance(cf_data, dict):
            return

        fcf_list = []

        # Look for yearly FCF data
        if 'yearly' in cf_data:
            for year_data in cf_data['yearly']:
                if isinstance(year_data, dict):
                    fcf_list.append({
                        'year': year_data.get('year'),
                        'fcf': year_data.get('fcf') or year_data.get('freeCashFlow')
                    })

        # Alternative structure
        if 'fcf' in cf_data:
            if isinstance(cf_data['fcf'], list):
                self.data['free_cash_flow'] = cf_data['fcf']
            else:
                fcf_list.append({'fcf': cf_data['fcf']})

        if fcf_list:
            self.data['free_cash_flow'] = fcf_list

    def _extract_valuation(self, soup) -> None:
        """Extract valuation metrics."""
        valuation = {}

        # Look for PE, PB, PEG ratios
        ratio_patterns = {
            'PE Ratio': re.compile(r'P/E|PE\s*Ratio', re.I),
            'PB Ratio': re.compile(r'P/B|PB\s*Ratio', re.I),
            'PEG Ratio': re.compile(r'PEG', re.I),
            'EV/EBITDA': re.compile(r'EV/EBITDA', re.I),
            'Dividend Yield': re.compile(r'Dividend\s*Yield', re.I)
        }

        for ratio_name, pattern in ratio_patterns.items():
            elem = soup.find(string=pattern)
            if elem:
                parent = elem.find_parent()
                if parent:
                    # Look for value in sibling or child elements
                    value_elem = parent.find_next(['span', 'div'], class_=re.compile(r'value|number'))
                    if value_elem:
                        valuation[ratio_name] = self.clean_number(value_elem.get_text(strip=True))

        # Also try to find in table format
        tables = soup.find_all('table')
        for table in tables:
            rows = table.find_all('tr')
            for row in rows:
                cells = row.find_all(['td', 'th'])
                if len(cells) >= 2:
                    label = cells[0].get_text(strip=True)
                    value = cells[1].get_text(strip=True)
                    for ratio_name, pattern in ratio_patterns.items():
                        if pattern.search(label):
                            valuation[ratio_name] = self.clean_number(value)

        self.data['valuation'] = valuation

    def _extract_growth(self, soup) -> None:
        """Extract growth metrics."""
        growth = {}

        growth_patterns = {
            'Revenue Growth': re.compile(r'Revenue\s*Growth|Sales\s*Growth', re.I),
            'Profit Growth': re.compile(r'Profit\s*Growth|PAT\s*Growth', re.I),
            'EPS Growth': re.compile(r'EPS\s*Growth', re.I),
            '3Y Revenue CAGR': re.compile(r'3.*Revenue.*CAGR|Revenue.*3.*CAGR', re.I),
            '5Y Revenue CAGR': re.compile(r'5.*Revenue.*CAGR|Revenue.*5.*CAGR', re.I),
            '10Y Revenue CAGR': re.compile(r'10.*Revenue.*CAGR|Revenue.*10.*CAGR', re.I)
        }

        for metric_name, pattern in growth_patterns.items():
            elem = soup.find(string=pattern)
            if elem:
                parent = elem.find_parent()
                if parent:
                    value_elem = parent.find_next(['span', 'div'], class_=re.compile(r'value|number'))
                    if value_elem:
                        growth[metric_name] = self.clean_number(value_elem.get_text(strip=True))

        self.data['growth'] = growth

    def _extract_cash_flow_data(self, soup, base_url: str) -> None:
        """Extract cash flow data, potentially from financials sub-page."""
        # Try to find cash flow section on main page
        cf_section = soup.find(['section', 'div'], id=re.compile(r'cash-flow', re.I))
        if cf_section:
            self._parse_cash_flow_table(cf_section)
            return

        # Try to fetch cash flow page if exists
        if '/stocks/' in base_url:
            cf_url = base_url.rstrip('/') + '/financials?checklist=basic&statement=cashflow'
            cf_soup = self.fetch_page(cf_url)
            if cf_soup:
                self._parse_cash_flow_page(cf_soup)

    def _parse_cash_flow_table(self, section) -> None:
        """Parse cash flow data from a table section."""
        table = section.find('table')
        if not table:
            return

        cash_flow_data = {}
        rows = table.find_all('tr')

        for row in rows:
            cells = row.find_all(['td', 'th'])
            if cells:
                metric = cells[0].get_text(strip=True).lower()

                if 'free cash flow' in metric or 'fcf' in metric:
                    fcf_values = []
                    for i, cell in enumerate(cells[1:]):
                        value = self.clean_number(cell.get_text(strip=True))
                        fcf_values.append(value)
                    cash_flow_data['fcf'] = fcf_values

                if 'operating' in metric and 'cash' in metric:
                    ocf_values = []
                    for cell in cells[1:]:
                        ocf_values.append(self.clean_number(cell.get_text(strip=True)))
                    cash_flow_data['operating_cash_flow'] = ocf_values

        self.data['cash_flow'] = cash_flow_data

    def _parse_cash_flow_page(self, soup) -> None:
        """Parse dedicated cash flow page."""
        # Find all tables
        tables = soup.find_all('table')
        for table in tables:
            rows = table.find_all('tr')
            for row in rows:
                cells = row.find_all(['td', 'th'])
                if cells and len(cells) > 1:
                    metric = cells[0].get_text(strip=True).lower()
                    if 'free cash flow' in metric or 'fcf' in metric:
                        fcf_list = []
                        for cell in cells[1:]:
                            value = self.clean_number(cell.get_text(strip=True))
                            fcf_list.append(value)
                        self.data['free_cash_flow'] = [{'fcf': v} for v in fcf_list if v is not None]

    def _is_numeric(self, value: str) -> bool:
        """Check if a string represents a numeric value."""
        cleaned = value.strip().replace(',', '').replace('%', '').replace('₹', '').replace('-', '')
        try:
            float(cleaned)
            return True
        except ValueError:
            return False

    def get_company_name(self) -> str:
        """Return the company name."""
        return self.company_name

    def get_free_cash_flow(self) -> List[Dict]:
        """Get free cash flow data."""
        return self.data.get('free_cash_flow', [])

    def get_peg_ratio(self) -> Optional[float]:
        """Get PEG ratio."""
        return self.data.get('valuation', {}).get('PEG Ratio')

    def get_price_history(self) -> Dict:
        """Get historical price data."""
        return self.data.get('price_history', {})
